src/type.c

```diff
@@ -53,33 +71,65 @@ static SEXP s3_type(SEXP x) {
   }
 }
 
+static SEXP vec_type_finalise_unspecified(SEXP x);
+static SEXP vec_type_finalise_dispatch(SEXP x);
 
 // [[ include("vctrs.h"); register() ]]
 SEXP vec_type_finalise(SEXP x) {
   if (x == R_NilValue) {
     return x;
   }
 
-  if (OBJECT(x)) {
-    if (vec_is_unspecified(x)) {
-      R_len_t n = Rf_length(x);
-      SEXP out = PROTECT(Rf_allocVector(LGLSXP, n));
-      r_lgl_fill(out, NA_LOGICAL, n);
-      UNPROTECT(1);
-      return out;
-    }
+  if (!OBJECT(x)) {
+    vec_assert(x, args_empty);
+    return x;
   }
 
-  if (!vec_is_partial(x)) {
-    vec_assert(x, args_empty);
+  if (vec_is_unspecified(x)) {
+    return vec_type_finalise_unspecified(x);
   }
 
-  switch (vec_typeof(x)) {
-  case vctrs_type_dataframe: return df_map(x, &vec_type_finalise);
-  case vctrs_type_s3:        return vctrs_dispatch1(syms_vec_type_finalise_dispatch, fns_vec_type_finalise_dispatch,
-                                                    syms_x, x);
-  default:                   return x;
+  if (vec_is_partial(x)) {
+    return vec_type_finalise_dispatch(x);
+  }
+
+  vec_assert(x, args_empty);
+
+  switch(class_type(x)) {
+  case vctrs_class_bare_tibble:
+  case vctrs_class_bare_data_frame:
+    return bare_df_map(x, &vec_type_finalise);
+
+  case vctrs_class_data_frame:
+    return df_map(x, &vec_type_finalise);
+
+  case vctrs_class_none:
+    Rf_errorcall(R_NilValue, "Internal error: Non-S3 classes should have returned by now");
+
+  default:
+    return vec_type_finalise_dispatch(x);
+  }
+}
+
+static SEXP vec_type_finalise_unspecified(SEXP x) {
```

Comment:
Use the new naming scheme `ptype`? This will reduce the amount of commit noise when we do rename old functions.

Reply:
Feel free to also rename functions as you touch them, if they are not used in too many places and the renaming causes little commit noise.